cbits/cbits.c

```diff
@@ -0,0 +1,56 @@
+#include <stdint.h>
+#include <string.h>
+#include <limits.h>
+#include <assert.h>
+
+/* Taken from:
+ *
+ * <http://stackoverflow.com/questions/776508/best-practices-for-circular-shift-rotate-operations-in-c>
+ *
+ * Intel is little-endian, we have to rotate right
+ */
+static inline uint32_t rotr32(uint32_t n, unsigned int c) {
+  const unsigned int mask = (CHAR_BIT*sizeof(n)-1);
+  c &= mask;  // avoid undef behaviour with NDEBUG.  0 overhead for most types / compilers
+  return (n>>c) | (n<<( (-c)&mask ));
+}
+
+/* - mask_shift is the initial shift of the mask.  It is specified in bytes.
+ */
+void _hs_mask_chunk(
+        uint32_t mask, int mask_shift,
+        uint8_t *payload_start, size_t payload_len,
+        uint8_t *target) {
+    const uint8_t *payload_end = payload_start + payload_len;
+
+    uint8_t *p = payload_start;
```

Comment:
While unaligned loads are allowed on common x86_64 and i386 architectures, they're generally discouraged since they can be slower on some processors (behind the scenes, two loads are performed).  So in this case, you need:

```
    while (p != payload_end && ((uintptr_t)p & 0x7)) {
        ...
    }
```

Annoyingly, we _also_ need to ensure that `target` is aligned as well.  I'm not sure if there's a way to do that easily in Haskell.  :-/